### PR TITLE
[NativeAOT-LLVM] Fix array allocators

### PR DESCRIPTION
**Pull request by SingleAccretion**
There were two bugs:
1) The overflow detection wasn't working properly due to C++ rules.
   The "align up" arithmetic was being performed in 32 bit integers.
2) When we overflow on 32 bit, we are supposed to OOM, not actually
   throw an overflow exception.

Fix both issues by simply delegating to the slow path helpers when we have large (65K+ elements) arrays.

With this fix, a small libraries test suite runs to completion on my machine:
```
Finished:    System.Collections.Specialized.Tests

=== TEST EXECUTION SUMMARY ===
Total: 2132, Errors: 0, Failed: 0, Skipped: 8, Time: 24.0773544s
```

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/nativeaot-llvm

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by jkotas** [comment]
It would be nice to upstream these changes to https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/portable.cpp

File: src/coreclr/nativeaot/Runtime/wasm/AllocFast.cpp [hunk 1]
```diff
@@ -242,36 +232,26 @@ COOP_PINVOKE_HELPER(Array*, RhpNewArrayAlign8, (void* pShadowStack, MethodTable*
...
     uint8_t* result = acontext->alloc_ptr;
     int requiresAlignObject = ((uint32_t)result) & 7;
     size_t paddedSize = size;
     if (requiresAlignObject)
     {
         if (paddedSize > SIZE_MAX - 12)
         {
-            ASSERT_UNCONDITIONALLY("NYI");  // TODO: Throw overflow
+            ThrowOverflowException(pShadowStack, pArrayEEType);
         }
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/wasm/AllocFast.cpp:255]: This check is dead code now that the large element sizes are handled as early out above.

**Pull request merged**